x/btcstaking/keeper/power_dist_change.go

```diff
@@ -106,21 +107,43 @@ func (k Keeper) recordVotingPowerAndCache(ctx context.Context, prevDc, newDc *ty
 	// set voting power table for each active finality providers at this height
 	for i := uint32(0); i < newDc.NumActiveFps; i++ {
 		fp := newDc.FinalityProviders[i]
-		k.SetVotingPower(ctx, fp.BtcPk.MustMarshal(), babylonTipHeight, fp.TotalVotingPower)
+		k.SetVotingPower(goCtx, fp.BtcPk.MustMarshal(), babylonTipHeight, fp.TotalVotingPower)
 	}
 
 	// find newly activated finality providers and execute the hooks by comparing
 	// the previous dist cache
-	newActivatedFinalityProviders := newDc.FindNewActiveFinalityProviders(prevDc)
-	for _, fp := range newActivatedFinalityProviders {
-		if err := k.hooks.AfterFinalityProviderActivated(ctx, fp.BtcPk); err != nil {
+	newActivatedFps := newDc.FindNewActiveFinalityProviders(prevDc)
+	for _, fp := range newActivatedFps {
+		if err := k.hooks.AfterFinalityProviderActivated(goCtx, fp.BtcPk); err != nil {
 			panic(fmt.Errorf("failed to execute after finality provider %s activated", fp.BtcPk.MarshalHex()))
 		}
-		k.Logger(sdk.UnwrapSDKContext(ctx)).Info("a new finality provider is activated", "pk", fp.BtcPk.MarshalHex())
+
+		statusChangeEvent := types.NewFinalityProviderStatusChangeEvent(fp.BtcPk, types.FinalityProviderStatus_STATUS_ACTIVE)
+		if err := sdkCtx.EventManager().EmitTypedEvent(statusChangeEvent); err != nil {
+			panic(fmt.Errorf(
+				"failed to emit FinalityProviderStatusChangeEvent with status %s: %w",
+				types.FinalityProviderStatus_STATUS_ACTIVE.String(), err))
+		}
+
+		k.Logger(sdkCtx).Info("a new finality provider becomes active", "pk", fp.BtcPk.MarshalHex())
+	}
```

Comment:
What do you think about moving the logic of emitting events out of `recordVotingPowerCache` as a separate function, and call this function in `UpdatePowerDist`? Suggesting this as `recordVotingPowerCache` is already a bit convoluted and has its own responsibility

Reply:
If it's outside of `recordVotingPowerCache`, we would need to do an extra iteration of getting active/inactive fps and getting new active/inactive ones, which slightly adds overhead. Do we want to do that?

Reply:
tbh I agree with Runchao that `recordVotingPowerCache` is already convoluted enough, but on the other hand I would not add additional extra iteration just for events sake. 

I am slightly leadning to:
1. merging this pr at is is to not block the api work and have all events in next version `v0.12.0`
2. schedule refactoring of this `power_dist_change.go` file and mechanism of updates for the next devnet3. It would also be great to tackle this `todo` while doing it -  https://github.com/babylonlabs-io/babylon/blob/v0.11.0/x/btcstaking/keeper/power_dist_change.go#L205

what do you guys think ?

Reply:
I also agree that we should separate `recordVotingPowerCache` into at least two functions: (1) transforming delegations into voting power with consideration of timestamping, jailing, etc. (2) emitting events and saving the cache.
Another refactoring point to consider: https://github.com/babylonlabs-io/babylon/issues/72

Reply:
>Another refactoring point to consider: https://github.com/babylonlabs-io/babylon/issues/72

uh this seems important 👍

Reply:
task created to track it https://github.com/babylonlabs-io/babylon/issues/145 (scheduled for devnet3)